tests/runner/Testcontracts.ml

```diff
@@ -102,22 +102,40 @@ let rec build_contract_tests_with_init_file env name exit_code i n
             "-libdir" :: (contract_dir ^/ lib_name) :: cur_args)
       in
       let args =
-        if disable_validate_json then "-disable-validate-json" :: args'
+        if disable_validate_json || env.server test_ctxt then
+          "-disable-validate-json" :: args'
         else args'
       in
-      let runner = "scilla-runner" in
+      (* Use scilla-client instead of scilla-runner when running tests in server-mode *)
+      let runner =
+        if env.server test_ctxt then "scilla-client" else "scilla-runner"
+      in
       print_cli_usage (env.print_cli test_ctxt) runner args;
       let test_name = name ^ "_" ^ istr in
       let goldoutput_file = dir ^/ "output_" ^ istr ^. "json" in
       let msg = cli_usage_on_err runner args in
+      let args =
+        if env.server test_ctxt then
+          [ "run"; "-argv"; String.concat args ~sep:" " ]
+        else args
+      in
       assert_command ~exit_code ~use_stderr:true ~ctxt:test_ctxt runner args
         ~foutput:(fun s ->
           (* if the test is supposed to succeed we read the output from a file,
                  otherwise we read from the output stream *)
           let interpreter_output =
-            if Poly.(exit_code = succ_code) then In_channel.read_all output_file
+            if Poly.(exit_code = succ_code) && not (env.server test_ctxt) then
+              In_channel.read_all output_file
             else BatStream.to_string s
           in
+          (* Extract the "result" field from the output of the scilla-server *)
+          let interpreter_output =
+            if env.server test_ctxt then
+              interpreter_output |> StateIPCTest.json_from_string
+              |> Yojson.Basic.Util.member "result"
```

Comment:
If there's any error, then the server will return a JSON result with `error` field sent and `result` not set. That needs to be checked here, otherwise it'll lead to an exception for "result" not being found.

It won't matter much but we'll have a tough time figuring out why the testsuite is failing. It's better if it's checked and the testsuite actually flags a proper `assert_failure` when that happens.

Reply:
Thanks for suggestion! Here is how I've fixed this:
1. From now on the `scilla-client` returns the contents of the `"result"` field or the `"error"` field, so there is no need to extract anything in the test suite
2. In case of error the `scilla-client` exits with code `1` and as you suggested above I added the call to `assert_failure` that raise an exception in such case (where exception string will be the output/error from the `scilla-client` executable):
```
if Poly.(exit_code = fail_code) && env.server test_ctxt then
  assert_failure interpreter_output;
```